src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -158,6 +158,12 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
             ty::Projection(projection) => (false, Some(projection)),
             _ => return,
         };
+        if self.in_const_context.get() {
```

Comment:
This is weirdly specific given that *anything* involving type parameters doesn't work.

Reply:
This is a _very_ common thing people encounter when trying to be too clever by half. I'm looking for other cases that are likely to be hit so that I can also handle them. One case I'm looking at now is

```rust
fn test<T: ?Sized>() {
    [0u8; std::mem::size_of::<&T>()];
}
```

Reply:
We should fix `generics_of` if the parent of `AnonConst` is an `Expr` (and probably any `Ty` nested in a body, but that's harder to check), I think I've suggested that before. It should just work, the query cycles only come from types outside bodies.

Reply:
And by "fix" I mean a trivial whitelist, much smaller than this PR.

Reply:
CC https://github.com/rust-lang/rust/pull/70452